log/DESIGN.md

```diff
@@ -0,0 +1,743 @@
+	// attributes.
+	front [5]KeyValue
+	nFront int // The number of attributes in front.
+	back []KeyValue
+}
+```
+
+[`Timestamp`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-timestamp)
+is accessed using following methods:
+
+```go
+func (r *Record) Timestamp() time.Time 
+func (r *Record) SetTimestamp(t time.Time) 
+```
+
+[`ObservedTimestamp`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-observedtimestamp)
+is accessed using following methods:
+
+```go
+func (r *Record) ObservedTimestamp() time.Time 
+func (r *Record) SetObservedTimestamp(t time.Time) 
+```
+
+[`SeverityNumber`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitynumber)
+is accessed using following methods:
+
+```go
+func (r *Record) Severity() Severity
+func (r *Record) SetSeverity(s Severity)
+```
+
+`Severity` type is defined and constants are based on
+[Displaying Severity recommendation](https://opentelemetry.io/docs/specs/otel/logs/data-model/#displaying-severity).
+Additionally, `Severity[Level]1` constants are defined to make the API more readable and user friendly.
+
+```go
+type Severity int
+
+const (
+	SeverityTrace1 Severity = 1 // TRACE
+	SeverityTrace2 Severity = 2 // TRACE2
+	SeverityTrace3 Severity = 3 // TRACE3
+	SeverityTrace4 Severity = 4 // TRACE4
+
+	SeverityDebug1 Severity = 5 // DEBUG
+	SeverityDebug2 Severity = 6 // DEBUG2
+	SeverityDebug3 Severity = 7 // DEBUG3
+	SeverityDebug4 Severity = 8 // DEBUG4
+
+	SeverityInfo1 Severity = 9  // INFO
+	SeverityInfo2 Severity = 10 // INFO2
+	SeverityInfo3 Severity = 11 // INFO3
+	SeverityInfo4 Severity = 12 // INFO4
+
+	SeverityWarn1 Severity = 13 // WARN
+	SeverityWarn2 Severity = 14 // WARN2
+	SeverityWarn3 Severity = 15 // WARN3
+	SeverityWarn4 Severity = 16 // WARN4
+
+	SeverityError1 Severity = 17 // ERROR
+	SeverityError2 Severity = 18 // ERROR2
+	SeverityError3 Severity = 19 // ERROR3
+	SeverityError4 Severity = 20 // ERROR4
+
+	SeverityFatal1 Severity = 21 // FATAL
+	SeverityFatal2 Severity = 22 // FATAL2
+	SeverityFatal3 Severity = 23 // FATAL3
+	SeverityFatal4 Severity = 24 // FATAL4
+
+	SeverityTrace = SeverityTrace1
+	SeverityDebug = SeverityDebug1
+	SeverityInfo  = SeverityInfo1
+	SeverityWarn  = SeverityWarn1
+	SeverityError = SeverityError1
+	SeverityFatal = SeverityFatal1
+)
+```
+
+[`SeverityText`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitytext)
+is accessed using following methods:
+
+```go
+func (r *Record) SeverityText() string
+func (r *Record) SetSeverityText(s string)
+```
+
+[`Body`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-body)
+is accessed using following methods:
+
+```go
+func (r *Record) Body() Value
+func (r *Record) SetBody(v Value)
+```
+
+[Log record attributes](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-attributes)
+are accessed using following methods:
+
+```go
+func (r *Record) WalkAttributes(f func(KeyValue) bool)
```

Comment:
You call them attributes in the method names, but the type is KeyValue. What about making the type Attr or Attribute?

Reply:
1. The name is the same as defined in attribute's package so I think it will be clear for people writing bridge implementations: https://pkg.go.dev/go.opentelemetry.io/otel/attribute#KeyValue. 
2. This type is also used for `Map`. This is why it is not called `Attribute` as there it is used to represent a "key-value".

Reply:
Calling `KeyValue` `Attr` here might help reduce confusion about us having a duplicate type. When they share the same name it becomes less clear what the difference between to two are and which one you are using.

Reply:
@MrAliasLike

```go
type KeyValue struct {
	Key   string
	Value Value
}

type Attr struct {
	Key   string
	Value Value
}
```

?

Reply:
I see one big issue with this proposal. I do not think we want to have factory functions for both `KeyValue` and `Attr` types. It would make the API surface unnecessary big and we may even have problem with naming the functions.

Reply:
I was thinking of just naming `KeyValue` `Attr` or `Attribute`, and having:

```go
func MapValue(kvs ...Attr) Value
```

Reply:
I don't have a strong preference on this. My concern is more about the questions:

- Can we disambiguate the difference between `attribute.KeyValue` and `log.KeyValue` by using different type names?
- Can we reduce potential misuse by users by naming `attribute.KeyValue` and `log.KeyValue` differently?

Reply:
I am opposed as for a map this is a real "key-value" and not an attribute. Attribute has semantical meaning.

Reply:
To @jba's point. New developers don't have the history we do with the `attribute.KeyValue` type. They may not intuitively know that `log.KeyValue` is an attribute in the OTel parlance.

Reply:
From offline conversation with @pellared:

- This package is expected to have ~90% of its use be by us OTel developers. We plan to implement bridges for the most popular logging libraries ourselves. Given we will all have the context needed to disambiguate these overlapping names developer confusion shouldn't be too big of an issue.
- For those bridges not developed by us, those developers will likely look at our existing bridges for inspiration. Our correct use of these types will be a reference to them.
- We can document the choice in the `KeyValue` type documentation. Specifically, we can state that these are a super-set of `attribute.KeyValue` and they are used for more than just attributes (i.e. log body types).

Based on these points, I'm in favor of keeping the existing `KeyValue` name.

Reply:
Documented as rejected proposal 75710811